# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/MockDragServiceController.cpp
# Commit: 384b025f56d3
# Full Hash: 384b025f56d3f87e045ecede6b1a16b71c0b5d4f
# Author: David P <daparks@mozilla.com>
# Date: 2024-07-04 16:09:52
# Regressor Bug: 1893119
# File Overlap Count: 1
# Description:
#   Bug 1893119: Part 21 - Separate nsIDragService and nsIDragSession implementations r=gstoll,geckoview-reviewers,rkraesig,win-reviewers,m_kato
#   
#   Split the class inheritance trees for nsIDragService and nsIDragSession.
#   Remember that, before the start of this patch series, the inheritance
#   diagram was:
# ==============================================================================

diff -r beeda5ef3712 -r 384b025f56d3 widget/MockDragServiceController.cpp
--- a/widget/MockDragServiceController.cpp	Thu Jul 04 07:48:11 2024 +0000
+++ b/widget/MockDragServiceController.cpp	Thu Jul 04 07:48:11 2024 +0000
@@ -17,8 +17,8 @@
 
 NS_IMPL_ISUPPORTS(MockDragServiceController, nsIMockDragServiceController)
 
-class MockDragService : public nsBaseDragService {
- public:
+class MockDragSession : public nsBaseDragSession {
+ protected:
   MOZ_CAN_RUN_SCRIPT nsresult
   InvokeDragSessionImpl(nsIWidget* aWidget, nsIArray* aTransferableArray,
                         const mozilla::Maybe<mozilla::CSSIntRegion>& aRegion,
@@ -50,12 +50,24 @@
     // Note that, like in non-mocked DND, we do this regardless of whether
     // the source and target were the same widget -- in that case,
     // EndDragSession is just called twice.
-    mDragAction = DRAGDROP_ACTION_MOVE;
-    StartDragSession(aWidget);
+    mDragAction = nsIDragService::DRAGDROP_ACTION_MOVE;
     return NS_OK;
   }
+};
 
-  bool IsMockService() override { return true; }
+class MockDragService : public nsBaseDragService {
+ public:
+  NS_IMETHOD GetIsMockService(bool* aRet) override {
+    *aRet = true;
+    return NS_OK;
+  }
+  uint32_t mLastModifierKeyState = 0;
+
+ protected:
+  already_AddRefed<nsIDragSession> CreateDragSession() override {
+    RefPtr<nsIDragSession> session = new MockDragSession();
+    return session.forget();
+  }
 };
 
 static void SetDragEndPointFromScreenPoint(